plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/CompoundProcessor.java

```diff
@@ -59,23 +62,32 @@ public String getType() {
 
     @Override
     public void execute(IngestDocument ingestDocument) throws Exception {
-        try {
-            for (Processor processor : processors) {
+        for (Processor processor : processors) {
+            try {
                 processor.execute(ingestDocument);
-            }
-        } catch (Exception e) {
-            if (onFailureProcessors.isEmpty()) {
-                throw e;
-            } else {
-                executeOnFailure(ingestDocument);
+            } catch (Exception e) {
+                if (onFailureProcessors.isEmpty()) {
+                    throw e;
+                } else {
+                    executeOnFailure(ingestDocument, e, processor.getType());
+                }
+                return;
             }
         }
+    }
+
+    private void resetErrorContext(Map<String, String> ingestMetadata) {
 
     }
 
-    void executeOnFailure(IngestDocument ingestDocument) throws Exception {
+    void executeOnFailure(IngestDocument ingestDocument, Exception cause, String failedProcessorType) throws Exception {
+        Map<String, String> ingestMetadata = ingestDocument.getIngestMetadata();
+        ingestMetadata.put(ON_FAILURE_MESSAGE_FIELD, cause.getMessage());
+        ingestMetadata.put(ON_FAILURE_PROCESSOR_FIELD, failedProcessorType);
```

Comment:
> The exception is only available in the context of the current on failure processors. They need to act upon it.

This answers my comment below about the need for removing too. I guess we have to document this behaviour well, sounds reasonable for now.